db/revision_cache_lru.go

```diff
@@ -272,42 +329,212 @@ func (rc *LRURevisionCache) getValue(docID, revID string, create bool) (value *r
+func (rc *LRURevisionCache) addToRevMapPostLoad(docID, revID string, cv *CurrentVersionVector) {
+	legacyKey := IDAndRev{DocID: docID, RevID: revID}
+	key := IDandCV{DocID: docID, Source: cv.SourceID, Version: cv.VersionCAS}
+
+	rc.lock.Lock()
+	defer rc.lock.Unlock()
+	// check for existing value in rev cache map (due to concurrent fetch by rev ID)
+	cvElem, cvFound := rc.hlvCache[key]
+	revElem, revFound := rc.cache[legacyKey]
+	if !cvFound {
+		// its possible the element has been evicted if we don't find the element above (high churn on rev cache)
+		// need to return doc revision to caller still but no need repopulate the cache
+		return
+	}
+	// Check if the element in the other lookup cache points to a value that has been updated in both lookup maps already
+	// This protects the following scenario:
+	//	- Process A is lookup via CV and Process B is lookup via revid, assume both process concurrent and both miss cache thus load from bucket
+	//	- Assume for this example both processes manage to 1. update their lookup map, 2. create a new value for the cache, 3. load from bucket to populate this value
+	//	- Now both processes need to update the other lookup map to their respective values
+	//	- Say process A acquires lock on rev cache for this function first, it looks up the key in the rev lookup map and finds it
+	//	assumes duplicate and removes the entry from the map and its respective element from the cache list
+	//	 - Process A continues on and updates the rev map now to point to its own value, process A then release the lock
+	//	 - Process B has been waiting on the lock, now its free it acquires the rev cache lock
+	//	 - Process B looks up key in cv lookup map, find and entry, assumes duplicate and removes the entry from the cv lookup map,
+	//	 and removes the element from teh cache list
+	//	 - We now find ourselves in a situation where both processes have removed both respective cache loads from the cache itself
+	//	 thus no entry in cache for this doc revision/version anymore. This needs to be avoided.
+	if revFound {
+		if cvElem == revElem {
+			// already match, return
+			return
+		}
+		rc.cache[legacyKey] = cvElem
+		rc.lruList.Remove(revElem)
+	}
+
+	// grab elem again using the hlv key in case the element in question has been evicted, in this case we don't want to update cache
+	if elem := rc.hlvCache[key]; elem != nil {
+		value := elem.Value.(*revCacheValue)
+		// we need to populate the value with the legacy key, so reacquire value lock here to populate the value with this
+		value.lock.Lock()
+		value.revID = revID
+		elem.Value = value
+		value.lock.Unlock()
+		rc.cache[legacyKey] = elem
+		for rc.lruList.Len() > int(rc.capacity) {
+			rc.purgeOldest_()
+		}
+	}
+}
+
+// addToHLVMapPostLoad will generate and entry in the CV lookup map for a new document entering the cache
+func (rc *LRURevisionCache) addToHLVMapPostLoad(docID, revID string, cv *CurrentVersionVector) {
```

Comment:
Same comments as addToRevMapPostLoad about less verbose comments, and not requiring the lower block of code.

Reply:
Good spot, I agree its not needed with this function. But the other function addToRevMapPostLoad still needs it in (see above comment for the context).